### PR TITLE
Reusable RS workspaces aka I don't know wtf am I doing :D

### PR DESCRIPTION
**Pull request by MiroCillik**

### CODE REVIEW

File: src/Docker/Runner/DataLoader/WorkspaceProviderFactoryFactory.php [hunk 1]
```diff
@@ -59,6 +60,43 @@ public function getWorkspaceProviderFactory(
...
+                $workspaces[1]['id'],
+                $configId,
+                $component->getId()
+            ));
+        }
+        return new ExistingFilesystemWorkspaceProviderFactory(
+            $this->workspacesApiClient,
+            $workspaceId,
+            $connectionString
+        );
```
Comment (pivnicek) [inline src/Docker/Runner/DataLoader/WorkspaceProviderFactoryFactory.php:97]: :-) we'll need to change this to `ExistingDatabaseWorkspaceProviderFactory`

Comment (pivnicek) [reply src/Docker/Runner/DataLoader/WorkspaceProviderFactoryFactory.php]: updatnuty

File: src/Docker/Runner/DataLoader/WorkspaceProviderFactoryFactory.php [hunk 2]
```diff
@@ -82,17 +82,21 @@ private function getWorkspaceFactoryForPersistentRedshiftWorkspace(Component $co
...
         } elseif (count($workspaces) === 1) {
             $workspaceId = $workspaces[0]['id'];
-            $password = $this->workspacesApiClient->resetWorkspacePassword($workspaceId)['connectionString'];
+            $password = $this->workspacesApiClient->resetWorkspacePassword($workspaceId)['password'];
             $this->logger->info(sprintf('Reusing persistent workspace "%s".', $workspaceId));
         } else {
-            throw new ApplicationException(sprintf(
-                'Multiple workspaces (total %s) found (IDs: %s, %s) for configuration "%s" of component "%s".',
+            $ids = array_column($workspaces, 'id');
+            sort($ids, SORT_NUMERIC);
```
Comment (pivnicek) [inline src/Docker/Runner/DataLoader/WorkspaceProviderFactoryFactory.php:89]: Shouldn't we rather `rsort` so we use the last created workspace?

File: Tests/backend-specific-tests/Redshift/DataLoaderPersistentRedshiftWorkspaceTest.php [hunk 3]
```diff
@@ -268,30 +350,40 @@ public function testRedshiftWorkspaceConfigMultipleWorkspace()
...
+        );
+        $dataLoader->storeOutput();
+
+        $this->assertTrue($logger->hasWarning(
+            sprintf(
+                'Multiple workspaces (total 2) found (IDs: %s) for configuration "%s" of component "%s", using "%s".',
+                implode(',', [$workspace1['id'], $workspace2['id']]),
+                $configurationId,
+                'keboola.runner-config-test',
+                $workspace1['id']
```
Comment (pivnicek) [inline Tests/backend-specific-tests/Redshift/DataLoaderPersistentRedshiftWorkspaceTest.php:369]: see `rsort` comment bellow.  I would have guessed this to be `$workspace2['id']`

**Review by pivnicek — COMMENTED** [review]
looks good, just a question about the sort order

**Review by pivnicek — COMMENTED** [review]
looks good, just a question about the sort order

**Review by pivnicek — APPROVED** [review]
great test 👍

**Review by ondrejhlavacek — APPROVED** [review]
lgtm

**Pull request merged**